routing/notifications.go

```diff
@@ -117,10 +117,12 @@ type topologyClient struct {
 // notifyTopologyChange notifies all registered clients of a new change in
 // graph topology in a non-blocking.
 func (r *ChannelRouter) notifyTopologyChange(topologyDiff *TopologyChange) {
-	r.RLock()
-	defer r.RUnlock()
+	// Get the clients inside a lock.
+	r.topologyClientMutex.RLock()
+	clients := r.topologyClients
```

Comment:
I know this protects the reference from concurrent access (as detected by the race detector).
But since maps are just pointers, this won't actually make sure that the data isn't changed while you loop over it. For that, I think you'd need to create a full copy of the map.
Or am I missing something?

Reply:
I agree that the decreased locking scope in `router.go` is sufficient.
But I think the loop over the `clients` here isn't sufficiently protected, since `clients` is indeed just a pointer to the same map.
See https://go.dev/play/p/Un2np9crRpJ, which I think is more or less what's happening here.

Reply:
oh yeah you are right! I misunderstood it and was thinking about the `clientID`, not the map. Indeed this won't work, now changed!